theano/gpuarray/subtensor.py

```diff
@@ -519,16 +519,24 @@ def perform(self, node, inputs, out_):
                 p += 1
                 narrays += 1
             else:
-                try:
-                    i.__index__()
-                    # We shift back the position of the array by the
-                    # number of dimensions that are removed by
-                    # indexing.  If ap is bigger than 0 it means we
-                    # have encountered at least one array.
-                    if ap >= 0:
-                        ap -= 1
-                except Exception:
-                    pass
+                if narrays == 0:
+                    try:
+                        i.__index__()
+                        # We shift back the position of the array by the
+                        # number of dimensions that are removed by
+                        # indexing.  If ap is bigger than 0 it means we
+                        # have encountered at least one array.
+                        if ap >= 0:
+                            ap -= 1
+                        # If this index is before the first array then
+                        # we will not move the array back to its
+                        # position.  Mark this by faking that there
+                        # are more than two arrays.  This is crazy
+                        # numpy behaviour so blame them.
+                        if narrays == 0:
```

Comment:
Why that new `if`? Could `narrays` have changed since the beginning of the block?